File: ios/Counter.m
//
//  Counter.m
//  CounterApp
//
//  Created by Joseph Kim on 2021/04/14.
//

#import <Foundation/Foundation.h>

#import "React/RCTBridgeModule.h"
#import "React/RCTEventEmitter.h"

// @interface RCT_EXTERN_MODULE(Counter, NSObject)
@interface RCT_EXTERN_MODULE(Counter, RCTEventEmitter)

RCT_EXTERN_METHOD(increment)
RCT_EXTERN_METHOD(getCount: (RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(
    decrement: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
@end

File: ios/Counter.swift
//
//  Counter.swift
//  CounterApp
//
//  Created by Joseph Kim on 2021/04/14.
//

import Foundation

@objc(Counter)
class Counter: RCTEventEmitter {
  
  private var count = 0
  
  @objc
  func increment() {
    count += 1
    print("**********IN SWIFT************")
    print("IN Swift : Count is \(count)")
    
    sendEvent(withName: "onIncrement", body: ["count": count])
  }
  
  override func supportedEvents() -> [String]! {
    return ["onIncrement"]
  }
    
  
  @objc
  func getCount(_ callback: RCTResponseSenderBlock) {
    callback([
      count,
      123.9,
      "third argument",
      [1,2.2,"3"],
      ["a": 1, "b": 2]
    ])
  }
  
  @objc
  func decrement(
    _ resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) -> Void {
    
    if (count == 0 ) {
      let error = NSError(domain: "", code: 200, userInfo: nil)
      reject("E_COUNT", "count cannot be negative",error)
    } else {
      count -= 1
      resolve("count was decremented")
    }
    
  }
  
  
  override func constantsToExport() -> [AnyHashable : Any]! {
      return [
        "number": 123.9,
        "string": "foo",
        "boolean": true,
        "array": [1, 22.2, "33"],
        "object": ["a":1, "b":2]
      ]
    }
  

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  }
